src/workerd/api/tests/http-socket-test.js

```diff
@@ -0,0 +1,501 @@
+  },
+};
+
+export const response404 = {
+  async test(ctrl, env, ctx) {
+    const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+    const httpClient = await internalNewHttpClient(socket);
+
+    const response = await httpClient.fetch('https://example.com/status/404');
+    assert.equal(response.status, 404);
+    const text = await response.text();
+    assert.equal(text, 'Not Found');
+  },
+};
+
+export const response500 = {
+  async test(ctrl, env, ctx) {
+    const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+    const httpClient = await internalNewHttpClient(socket);
+
+    const response = await httpClient.fetch('https://example.com/status/500');
+    assert.equal(response.status, 500);
+    const text = await response.text();
+    assert.equal(text, 'Internal Server Error');
+  },
+};
+
+export const redirect301 = {
+  async test(ctrl, env, ctx) {
+    const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+    const httpClient = await internalNewHttpClient(socket);
+
+    // TODO(cleanup) when enabling multiple fetches we can only then do redirects.
+    await assert.rejects(httpClient.fetch('https://example.com/redirect'), {
+      name: 'Error',
+      message:
+        'Fetcher created from internalNewHttpClient can only be used once',
+    });
+    /*
+    assert.equal(response.status, 200);
+    const text = await response.text();
+    assert.equal(text, 'pong');
+    */
+  },
+};
+
+export const multipleRequests = {
+  async test(ctrl, env, ctx) {
+    const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+    const httpClient = await internalNewHttpClient(socket);
+
+    // First request
+    const response1 = await httpClient.fetch('https://example.com/ping');
+    assert.equal(response1.status, 200);
+    const text1 = await response1.text();
+    assert.equal(text1, 'pong');
+
+    // Second request on same connection
+    // TODO(cleanup) we want this to fail during this initial implementation but later we should
+    // support mutiple fetches (perhaps behind a compat flag)
+    await assert.rejects(httpClient.fetch('https://example.com/json'), {
+      name: 'Error',
+      message:
+        'Fetcher created from internalNewHttpClient can only be used once',
+    });
+    /*
+     * assert.equal(response2.status, 200);
+     * const data = await response2.json();
+     * assert.deepEqual(data, { message: 'Hello from HTTP socket server' });
+     */
+  },
+};
+
+export const multipleConcurrentRequests = {
+  async test(ctrl, env, ctx) {
+    const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+    const httpClient = await internalNewHttpClient(socket);
+
+    // TODO(cleanup) when multiple fetches are enabled make sure this message is changed
+    await assert.rejects(
+      Promise.all([
+        httpClient.fetch('https://example.com/ping'),
+        httpClient.fetch('https://example.com/json'),
```

Comment:
Once we update it to allow the fetcher to be reused, this would still be a failure if the fetches are expected to be active at the same time. If the fetches are queued then this should work so long as the second fetch is not actually initiated until the second completes.